pkg/filter/filter.go

```diff
@@ -108,24 +119,41 @@ func JoinResourceTypeMatchers(ms ...ResourceTypeMatcher) ResourceTypeMatcher {
 
 // ResourceRequirement allows specifying a resource requirement by type and/or name
 type ResourceRequirement struct {
+	// Provides the Name of the resource object
 	v1.LocalObjectReference `json:",inline,omitempty"`
+	// Provides the Group, Version, and Resource values (GVR)
 	ResourceTypeRequirement `json:",inline,omitempty"`
+	// Specifies a set of label requirements to be used as filters for matches
+	metav1.LabelSelector `json:",inline,omitempty"`
 }
 
-// Matches returns true if the specified resource name/GVR matches the requirement
-func (r ResourceRequirement) Matches(name string, gvr schema.GroupVersionResource) bool {
+// Matches returns true if the specified resource name/GVR/labels matches the requirement
+func (r ResourceRequirement) Matches(name string, gvr schema.GroupVersionResource, resourceLabels map[string]string) bool {
 	// If the requirement does not specify a resource name - only check the
 	// ResourceTypeRequirement i.e. GVR match
-	if r.LocalObjectReference.Name == "" {
-		return r.ResourceTypeRequirement.Matches(gvr)
+	if !(matches(r.Name, name) && r.ResourceTypeRequirement.Matches(gvr)) {
+		return false
 	}
-	return matches(r.Name, name) && r.ResourceTypeRequirement.Matches(gvr)
+	if len(resourceLabels) == 0 || (len(r.MatchExpressions) == 0 && len(r.MatchLabels) == 0) {
+		// If there are no resourceLabels to be matched, the LabelSelector is ignored.
+		// If there is no LabelSelector defined, the specified resourceLabels are ignored.
+		return true
+	}
+	sel, err := metav1.LabelSelectorAsSelector(&r.LabelSelector)
+	if err != nil {
+		// A match was found on Name or GVR but labels could not be evaluated.
```

Comment:
Arguably we should have validated the label when policy or action was validated (let's make sure to to so), but if we get here we have a policy question of whether to ignore the label and treat as match or fail to match.  My sense is that we shouldn't match (better to do nothing than do the wrong thing), but if there's precedent elsewhere I'd opt for consistency.  Do you know of any analogues?  What's your reasoning?

Reply:
Reasoning is just a gut feel. I'm more inclined to return the error rather than burying it and making an assumption either way.

Reply:
Additional tests cleared up and changes made that align with this comment.